art/attacks/poisoning/feature_collision_attack.py

```diff
@@ -0,0 +1,279 @@
+        :type classifier: (`art.estimators.NeuralNetworkMixin`, `art.estimators.BaseEstimator`)
+        :param target: The target input to misclassify at test time
+        :type target: `np.ndarray`
+        :param feature_layer: The name of the feature representation layer
+        :type feature_layer: `str`
+        :param learning_rate: The learning rate of clean-label attack optimization
+        :type learning_rate: `float`
+        :param decay_coeff: The decay coefficient of the learning rate
+        :type decay_coeff: `float`
+        :param stopping_tol: The tolerance for relative change in objective function
+        :type stopping_tol: `float`
+        :param num_old_obj: The number of old objective values to store
+        :type num_old_obj: `int`
+        :param max_iter: The maximum number of iterations for the attack
+        :type max_iter: `int`
+        :param similarity_coeff: The maximum number of iterations for the attack
+        :type similarity_coeff: `float`
+        :param watermark: Whether The opacity of the watermarked target image
+        :type watermark: `float`
+        :param kwargs: Extra optional keyword arguments
+        """
+        super().__init__(classifier)
+
+        if not isinstance(classifier, (NeuralNetworkMixin, BaseEstimator)):
+            raise TypeError("Classifier must be a neural network")
+
+        self.classifier = classifier
+        self.target = target
+        self.feature_layer = feature_layer
+        self.learning_rate = learning_rate
+        self.decay_coeff = decay_coeff
+        self.stopping_tol = stopping_tol
+        self.num_old_obj = num_old_obj
+        self.max_iter = max_iter
+        self.similarity_coeff = similarity_coeff
+        self.watermark = watermark
+
+        self.set_params(**kwargs)
+
+    def poison(self, x, y=None, **kwargs):
+        """
+        Iteratively finds optimal attack points starting at values at x
+
+        :param x: The base images to begin the poison process
+        :type x: `np.ndarray`
+        :param y: Not used  in this attack (clean-label)
+        :return: An tuple holding the (poisoning examples, poisoning labels).
+        :rtype: `(np.ndarray, np.ndarray)`
+        """
+
+        num_poison = len(x)
+        final_attacks = []
+        if num_poison == 0:
+            raise ValueError("Must input at least one poison point")
+
+        # TODO: ensure class of x does not match class of target
+        target_features = self.classifier.get_activations(self.target, self.feature_layer, 1)
+        for init_attack in x:
+            old_attack = np.expand_dims(np.copy(init_attack), axis=0)
+            poison_features = self.classifier.get_activations(old_attack, self.feature_layer, 1)
+            old_objective = self.objective(poison_features, target_features, init_attack, old_attack)
+            last_m_objectives = [old_objective]
+            # TODO: change to while with convergence (add convergence params)
```

Comment:
Please add convergence check.

Reply:
I currently have a convergence check for stopping after changes to the poison image are too small. Authors of the original paper also added a convergence check based on the objective value (changes with every problem). Was wondering what your thoughts were on adding this extra check

Reply:
How about making the choice of convergence criteria an option for `FeatureCollisionAttack` through an additional argument? That way the user could decide.

Reply:
added in new commit